File: RadiantSDK/TrueTest/Visual Studio 2013/C#/Device Interface Standard/StandardDI.cs
using System.Windows.Forms;
using Microsoft.VisualBasic;
using System.Collections.Generic;
using System;
using System.IO;
using System.ComponentModel;
using System.IO.Ports;
using System.Xml.Serialization;
using TrueTestPatternGenerator;
using TrueTestEngine;
using System.Runtime.CompilerServices;

namespace StandardDI
{
    [Serializable()]
    public class StandardDI : PatternGeneratorBase, IDisposable
    {
        private int NumberOfSequenceExposures = 0;
        private int ExposuresCompleted = 0;
        private bool AckRecieved = false;

        // Enumerator list for the avialble commands
        public enum SerialCommand
        {
            SerialNumber,
            SerialAcknowledge,
            DeviceReady,
            ShowPattern,
            PatternAcknowledge,
            Unload,
            UnloadAcknowledge,
            Result,
            ResultAcknowledge
        }

        // Enumerator list which represents the termination character of a signal
        public enum TerminatorEnum
        {
            None = 0,
            Cr = 1,
            Lf = 2,
            CrLf = 3,
            LfCr = 4
        }

        [XmlIgnore()]
        [NonSerialized()]
        [Browsable(false)]
        private SerialPort _SerialPort;

        [XmlIgnore()]
        [Browsable(false)]
        public SerialPort SerialPort
        {
            [MethodImpl(MethodImplOptions.Synchronized)]
            get
            {
                return _SerialPort;
            }

            [MethodImpl(MethodImplOptions.Synchronized)]
            set
            {
                if (_SerialPort != null)
                {
                    _SerialPort.DataReceived -= SerialPort_DataReceived;
                }

                _SerialPort = value;
                if (_SerialPort != null)
                {
                    _SerialPort.DataReceived += SerialPort_DataReceived;
                }
            }
        }

        [Category("Fixture Serial Port Parameters")]
        [Browsable(true)]
        [Description("The serial port name used to communicate with the fixture PLC.")]
        public string PortName { get; set; }

        [Category("Fixture Serial Port Parameters")]
        [Browsable(true)]
        [Description("Parity of communication protocol.")]
        public System.IO.Ports.Parity Parity { get; set; }

        [Category("Fixture Serial Port Parameters")]
        [Browsable(true)]
        [Description("Number of databits in communication protocol.")]
        public int DataBits { get; set; }

        [Category("Fixture Serial Port Parameters")]
        [Browsable(true)]
        [Description("Stop Bits in communication protocol.")]
        public StopBits StopBits { get; set; }

        [Category("Fixture Serial Port Parameters")]
        [Browsable(true)]
        [Description("The Baud Rate of communication protocol.")]
        public int BaudRate { get; set; }

        [Category("Fixture Serial Port Parameters")]
        [Browsable(true)]
        [Description("The number of miliseconds to wait before considering communcation dropped.")]
        public int WaitTime { get; set; }

        private string mTerminator = Constants.vbCrLf;
        [Category("Fixture Serial Port Parameters")]
        [Browsable(true)]
        [Description("The terminating character in communication protocol.")]
        public TerminatorEnum Terminator
        {
            get
            {
                switch (mTerminator)
                {
                    case "":
                        {
                            return TerminatorEnum.None;
                        }

                    case Constants.vbCr:
                        {
                            return TerminatorEnum.Cr;
                        }

                    case Constants.vbLf:
                        {
                            return TerminatorEnum.Lf;
                        }

                    case Constants.vbCrLf:
                        {
                            return TerminatorEnum.CrLf;
                        }

                    case Constants.vbLf + Constants.vbCr:
                        {
                            return TerminatorEnum.LfCr;
                        }

                    default:
                        {
                            return TerminatorEnum.CrLf;
                        }
                }
            }
            set
            {
                switch (value)
                {
                    case TerminatorEnum.None:
                        {
                            mTerminator = "";
                            break;
                        }

                    case TerminatorEnum.Cr:
                        {
                            mTerminator = Constants.vbCr;
                            break;
                        }

                    case TerminatorEnum.Lf:
                        {
                            mTerminator = Constants.vbLf;
                            break;
                        }

                    case TerminatorEnum.CrLf:
                        {
                            mTerminator = Constants.vbCrLf;
                            break;
                        }

                    case TerminatorEnum.LfCr:
                        {
                            mTerminator = Constants.vbLf + Constants.vbCr;
                            break;
                        }

                    default:
                        {
                            mTerminator = Constants.vbCrLf;
                            break;
                        }
                }
            }
        }

        internal bool Initialized = false;



        [Category("Log File Parameters")]
        [Browsable(true)]
        [Description("The file path of the log file")]
        public string LogFilePath { get; set; }

        [Category("Log File Parameters")]
        [Browsable(true)]
        [Description("Determines if a communications log will be written>")]
        public bool WriteToCommunicationsLog { get; set; }


        //Set initial defaults
        StandardDI()
        {
            SerialPort = null;
            PortName = "COM1";
            Parity = Parity.None;
            DataBits = 8;
            StopBits = StopBits.One;
            BaudRate = 9600;
            WaitTime = 3000;
            LogFilePath = @"C:\Radiant Vision Systems Data\TrueTest\AppData";
            WriteToCommunicationsLog = true;
        }


        /// <summary>
        /// Overwrites Initialize() in PatternGeneraorBase so the serial port will be initialized along with the PG.
        /// </summary>
        protected override bool Initialize()
        {
            Initialized = false;
            AddEventHandlers();
            return InitializeSerialPort();
        }

        internal void AddEventHandlers()
        {
            TrueTest.ExposureComplete += ExposureComplete;
            TrueTest.SequenceComplete += SequenceComplete;
            TrueTest.SequenceRunAllStarted += PrepareForSequenceRunAll;
        }

        internal void RemoveHandlers()
        {
            TrueTest.ExposureComplete -= ExposureComplete;
            TrueTest.SequenceComplete -= SequenceComplete;
            TrueTest.SequenceRunAllStarted -= PrepareForSequenceRunAll;
        }

        /// <summary>
        /// Initializes the serial port.
        /// </summary>
        /// <returns> Returns true if port was successfully initialized.</returns>
        private bool InitializeSerialPort()
        {
            try
            {
                SerialPort = new SerialPort(PortName, BaudRate, Parity, DataBits, StopBits)
                {
                    ReceivedBytesThreshold = 1,
                    DiscardNull = true,
                    NewLine = mTerminator,
                    WriteTimeout = WaitTime,
                    ReadTimeout = WaitTime
                };
                SerialPort.Open();
                Initialized = true;

                // Initialization for pattern generator
                this.IsInitialized = true;
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message, "Fixture Serial Port Error");
                SerialPort = null;
            }

            return Initialized;
        }

        protected override bool IsInitializationRequired(TrueTestPatternGenerator.PatternGeneratorBase NewPGObject)
        {
            if (NewPGObject == null)
                return false;
            return true; // reinitialize always
        }

        /// <summary>
        /// Overwrites ShutDown() in PatternGeneraorBase so that the serial port will be closed along with the PG.
        /// </summary>
        /// <returns> true </returns>
        protected override bool ShutDown()
        {
            if (Initialized == false)
                return true;
            Close();
            RemoveHandlers();
            this.IsInitialized = false;
            return true;
        }

        /// <summary>
        /// Closes serial port.
        /// </summary>
        private void Close()
        {
            try
            {
                if (SerialPort != null)
                {
                    if (SerialPort.IsOpen)
                    {
                        SerialPort.Close();
                        Initialized = false;
                    }
                }
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message, "Fixture Serial Port Error");
            }
        }


        /// <summary>
        /// Listens for commands from serial port and calls functions depending upon command received.
        /// </summary>
        private void SerialPort_DataReceived(object sender, System.IO.Ports.SerialDataReceivedEventArgs e)
        {
            try
            {
                // 20ms Sleep introduced in order to allow the buffer to completely fill.
                System.Threading.Thread.Sleep(20);

                string RecievedString = SerialPort.ReadLine();
                PacketSerial ReceivedPacket = new PacketSerial(RecievedString);

                switch (ReceivedPacket.Request())
                {
                    case SerialCommand.SerialNumber:
                        {
                            string SerialNumber = ReceivedPacket.Data[0];
                            WriteToLog("Received: SN," + SerialNumber);
                            TrueTest.set_SerialNumber(0, SerialNumber);
                            SerialPort.WriteLine("SA," + SerialNumber);
                            return;
                        }

                    case SerialCommand.DeviceReady:
                        {
                            WriteToLog("Received: RD");

                            // Run the sequence
                            TrueTest.SequenceRunAll();

                            return;
                        }

                    case SerialCommand.PatternAcknowledge:
                        {
                            WriteToLog("Received: PA," + ReceivedPacket.Data[0]);
                            AckRecieved = true;
                            break;
                        }

                    case SerialCommand.UnloadAcknowledge:
                        {
                            WriteToLog("Received: UA");
                            break;
                        }

                    case SerialCommand.ResultAcknowledge:
                        {
                            WriteToLog("Received: RA," + ReceivedPacket.Data[0]);
                            break;
                        }

                    default:
                        {
                            WriteToLog("Unrecognized command: " + RecievedString);
                            break;
                        }
                }
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message, "SerialPort Error");
                WriteToLog("Receive Error: " + ex.Message);
            }
        }



        private void ExposureComplete(object sender, ExposureCompleteEventArgs e)
        {
            ExposuresCompleted += 1;
            WriteToLog("Exposure Number " + ExposuresCompleted + " is complete.");
            if (ExposuresCompleted >= NumberOfSequenceExposures)
            {
                // We've taken all of the exposures and we can unload the panel now
                PacketSerial UnloadPacket = new PacketSerial(SerialCommand.Unload);
                UnloadPacket.Send(this);
            }
        }

        public void MeasurementComplete(object sender, MeasurementCompleteEventArgs e)
        {
            WriteToLog("Measurement Complete...");
        }

        public void AnalysisComplete(object sender, AnalysisCompleteEventArgs e)
        {
            WriteToLog("Analysis finished...");
        }

        public void SequenceComplete(object sender, SequenceCompleteEventsArgs e)
        {
            WriteToLog("Sequence Complete...");

            string ResultString = e.PassFail == TrueTest.AnalysisResultEnum.Pass ? "OK" : "NG";

            List<string> DataStringList = new List<string>() { ResultString };

            PacketSerial ResultPacket = new PacketSerial(SerialCommand.Result, DataStringList);
            ResultPacket.Send(this);
        }


        /// <summary>
        /// This is a blocking call that doesn't finish until the requested pattern is confirmed to be displayed.
        /// </summary>
        /// <param name="p"> The pattern to be displayed and confirmed. </param>
        protected override void ShowPattern(TrueTestPattern p)
        {
            if (p == null)
                return;

            // Check that the pattern is the appropriate type for the DI to send.
            if (!p.Pattern.GetType().Equals(typeof(SimplePattern)))
                return;

            AckRecieved = false;
            SimplePattern pattern = (SimplePattern)p.Pattern;
            string PatternNumber = pattern.ImageNumber.ToString();
            List<string> PacketDataList = new List<string>() { PatternNumber };
            PacketSerial PatternPacket = new PacketSerial(SerialCommand.ShowPattern, PacketDataList);
            PatternPacket.Send(this);

            if (!WaitForAck())
            {
                // Fixture timed out changing the pattern
                WriteToLog("Fixture did not send the Pattern Ready (PR) command after TrueTest requested pattern #" + PatternNumber + ".");
                TrueTest.SequenceStop(); // Cancel sequence
            }
        }

        private bool WaitForAck()
        {
            // Wait for ack from board

            // loop in increments of 10ms
            int Timeout = WaitTime / 10;
            int TimeoutCounter = 0;
            do
            {
                System.Threading.Thread.Sleep(10);
                TimeoutCounter += 1;
            }
            while (!(AckRecieved | TimeoutCounter > Timeout));
            if (TimeoutCounter >= Timeout)
                // never got anything back
                WriteToLog("No acknowledgement from board. Please check connection and power.");

            if (AckRecieved)
            {
                AckRecieved = false;
                return true;
            }

            AckRecieved = false;
            return AckRecieved;
        }

        /// <summary>
        /// Writes the string to the com log with a time stamp.
        /// </summary>
        private object LogSyncLock = new object();
        internal void WriteToLog(string Message)
        {
            DateTime dt = DateTime.Now;
            string LogFilename = LogFilePath + @"\" + dt.ToString("yyyyMMdd") + " Comm Log.txt";
            bool FileExists = File.Exists(LogFilename);

            try
            {
                lock (LogSyncLock)
                {
                    StreamWriter sw = new StreamWriter(LogFilename, true);

                    if (FileExists == false)
                        sw.WriteLine("Log created" + Constants.vbTab + dt.ToString("yyyy/MM/dd HH:mm:ss.fff"));

                    if (Message != null)
                    {
                        Message = Message + Constants.vbTab + dt.ToString("yyyy/MM/dd HH:mm:ss.fff");
                        sw.WriteLine(Message);
                    }

                    sw.Close();
                }
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message, "Log File Error");
            }
        }


        /// <summary>
        /// Checks the sequence for each unique measurement as analyses may reuse measurements already taken for others.
        /// </summary>
        public void PrepareForSequenceRunAll(object sender, EventArgs e)
        {
            // Calculate how many exposures are going to be required to run the sequence
            List<PatternSetup> PatternList = new List<PatternSetup>();
            NumberOfSequenceExposures = 0;
            var sequence = TrueTest.get_Sequence(0);
            foreach (TrueTestEngine.SequenceItem SeqStep in sequence.Items)
            {
                if (!SeqStep.Selected)
                    continue;
                PatternSetup ps = sequence.GetPatternSetupByName(SeqStep.PatternSetupName);
                if (ps == null)
                    continue;

                // Check to see if we already counted this pattern's exposures
                bool RepeatPattern = false;
                foreach (PatternSetup P in PatternList)
                {
                    if (P == ps)
                    {
                        RepeatPattern = true;
                        break;
                    }
                }

                if (!RepeatPattern)
                {
                    PatternList.Add(ps);
                    NumberOfSequenceExposures += 1;
                }
            }
            WriteToLog("This sequence will require " + NumberOfSequenceExposures + " camera exposures.");
            // Reset the number of exposures completed
            ExposuresCompleted = 0;
        }
    }
}

File: RadiantSDK/TrueTest/Visual Studio 2013/VB.NET/Device Interface Standard/StandardDI.vb
Imports System.IO
Imports System.ComponentModel
Imports System.IO.Ports
Imports System.Xml.Serialization
Imports TrueTestPatternGenerator
Imports TrueTestEngine

<Serializable()>
Public Class StandardDI

    'Need to inheirt PatternGeneratorBase so TrueTest will recognize this as a device interface. 
    Inherits PatternGeneratorBase
    Implements IDisposable
    Private SequenceIsRunning As Boolean = False
    Private CommandDelivered As Boolean = False
    Private NumberOfSequenceExposures As Integer = 0
    Private ExposuresCompleted As Integer = 0
    Private AckRecieved As Boolean = False
    Private RequestedPattern As Integer

    'Enumerator list for the avialble commands
    Enum SerialCommand
        SerialNumber
        SerialAcknowledge
        DeviceReady
        ShowPattern
        PatternAcknowledge
        Unload
        UnloadAcknowledge
        Result
        ResultAcknowledge
    End Enum

    'Enumerator list which represents the termination character of a signal
    Public Enum TerminatorEnum
        None = 0
        Cr = 1
        Lf = 2
        CrLf = 3
        LfCr = 4
    End Enum

#Region "Fixture Serial Port Parameters"
    <XmlIgnore(), NonSerialized(), Browsable(False)>
    Friend WithEvents SerialPort As SerialPort = Nothing

    <CategoryAttribute("Fixture Serial Port Parameters"), BrowsableAttribute(True),
    DescriptionAttribute("The serial port name used to communicate with the fixture PLC.")>
    Public Property PortName As String = "COM1"

    <CategoryAttribute("Fixture Serial Port Parameters"), BrowsableAttribute(True),
    DescriptionAttribute("Parity of communication protocol.")>
    Public Property Parity As IO.Ports.Parity = IO.Ports.Parity.None

    <CategoryAttribute("Fixture Serial Port Parameters"), BrowsableAttribute(True),
    DescriptionAttribute("Number of databits in communication protocol.")>
    Public Property DataBits As Integer = 8

    <CategoryAttribute("Fixture Serial Port Parameters"), BrowsableAttribute(True),
    DescriptionAttribute("Stop Bits in communication protocol.")>
    Public Property StopBits As StopBits = StopBits.One

    <CategoryAttribute("Fixture Serial Port Parameters"), BrowsableAttribute(True),
    DescriptionAttribute("The Baud Rate of communication protocol.")>
    Public Property BaudRate As Integer = 9600

    <CategoryAttribute("Fixture Serial Port Parameters"), BrowsableAttribute(True),
    DescriptionAttribute("The number of miliseconds to wait before considering communcation dropped.")>
    Public Property WaitTime As Integer = 3000

    Private mTerminator As String = vbCrLf
    <CategoryAttribute("Fixture Serial Port Parameters"), BrowsableAttribute(True),
    DescriptionAttribute("The terminating character in communication protocol.")>
    Public Property Terminator As TerminatorEnum
        Get
            Select Case mTerminator
                Case ""
                    Return TerminatorEnum.None
                Case vbCr
                    Return TerminatorEnum.Cr
                Case vbLf
                    Return TerminatorEnum.Lf
                Case vbCrLf
                    Return TerminatorEnum.CrLf
                Case vbLf & vbCr
                    Return TerminatorEnum.LfCr
                Case Else
                    Return TerminatorEnum.CrLf
            End Select
        End Get
        Set(ByVal term As TerminatorEnum)
            Select Case term
                Case TerminatorEnum.None
                    mTerminator = ""
                Case TerminatorEnum.Cr
                    mTerminator = vbCr
                Case TerminatorEnum.Lf
                    mTerminator = vbLf
                Case TerminatorEnum.CrLf
                    mTerminator = vbCrLf
                Case TerminatorEnum.LfCr
                    mTerminator = vbLf & vbCr
                Case Else
                    mTerminator = vbCrLf
            End Select
        End Set
    End Property

    Friend Initialized As Boolean = False

#End Region

#Region "Log File Communication"

    <CategoryAttribute("Log File Parameters"), BrowsableAttribute(True),
    DescriptionAttribute("The file path of the log file")>
    Public Property LogFilePath As String = "C:\Radiant Vision Systems Data\TrueTest\AppData"

    <CategoryAttribute("Log File Parameters"), BrowsableAttribute(True),
    DescriptionAttribute("Determines if a communications log will be written>")>
    Public Property WriteToCommunicationsLog As Boolean = True
#End Region

#Region "Initialization and ShutDown"
    ''' <summary>
    ''' Overwrites Initialize() in PatternGeneraorBase so the serial port will be initialized along with the PG.
    ''' </summary>
    Protected Overrides Function Initialize() As Boolean
        Initialized = False
        AddEventHandlers()
        Return InitializeSerialPort()
    End Function

    Friend Sub AddEventHandlers()
        AddHandler TrueTest.ExposureComplete, AddressOf ExposureComplete
        AddHandler TrueTest.SequenceComplete, AddressOf SequenceComplete
        AddHandler TrueTest.SequenceRunAllStarted, AddressOf PrepareForSequenceRunAll
    End Sub

    Friend Sub RemoveHandlers()
        RemoveHandler TrueTest.ExposureComplete, AddressOf ExposureComplete
        RemoveHandler TrueTest.SequenceComplete, AddressOf SequenceComplete
        RemoveHandler TrueTest.SequenceRunAllStarted, AddressOf PrepareForSequenceRunAll
    End Sub

    ''' <summary>
    ''' Initializes the serial port.
    ''' </summary>
    ''' <returns> Returns true if port was successfully initialized.</returns>
    Private Function InitializeSerialPort() As Boolean
        Try
            SerialPort = New SerialPort(PortName, BaudRate, Parity, DataBits, StopBits) With {
                .ReceivedBytesThreshold = 1,
                .DiscardNull = True,
                .NewLine = mTerminator,
                .WriteTimeout = WaitTime,
                .ReadTimeout = WaitTime
            }
            SerialPort.Open()
            Initialized = True

            'Initialization for pattern generator
            Me.IsInitialized = True

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Fixture Serial Port Error")
            SerialPort = Nothing

        End Try

        Return Initialized
    End Function

    Protected Overrides Function IsInitializationRequired(ByVal NewPGObject As TrueTestPatternGenerator.PatternGeneratorBase) As Boolean
        If NewPGObject Is Nothing Then Return False
        Return True 'reinitialize always
    End Function

    ''' <summary>
    ''' Overwrites ShutDown() in PatternGeneraorBase so that the serial port will be closed along with the PG.
    ''' </summary>
    ''' <returns> true </returns>
    Protected Overrides Function ShutDown() As Boolean

        If Initialized = False Then Return True
        Close()
        RemoveHandlers()
        Me.IsInitialized = False
        Return True

    End Function

    ''' <summary>
    ''' Closes serial port.
    ''' </summary>
    Private Sub Close()
        Try
            If SerialPort IsNot Nothing Then
                If SerialPort.IsOpen Then
                    SerialPort.Close()
                    Initialized = False
                End If
            End If
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Fixture Serial Port Error")
        End Try

    End Sub
#End Region

#Region "Serial Port Send/Receive Functions"

    ''' <summary>
    ''' Listens for commands from serial port and calls functions depending upon command received.
    ''' </summary>
    Private Sub SerialPort_DataReceived(sender As Object, e As System.IO.Ports.SerialDataReceivedEventArgs) Handles SerialPort.DataReceived
        Try
            '20ms Sleep introduced in order to allow the buffer to completely fill.
            Threading.Thread.Sleep(20)

            Dim RecievedString As String = SerialPort.ReadLine()
            Dim ReceivedPacket As New PacketSerial(RecievedString)

            Select Case ReceivedPacket.Request
                Case SerialCommand.SerialNumber
                    Dim SerialNumber As String = ReceivedPacket.Data(0)
                    WriteToLog("Received: SN," & SerialNumber)
                    TrueTest.SerialNumber = SerialNumber
                    SerialPort.WriteLine("SA," & SerialNumber)
                    Exit Sub
                Case SerialCommand.DeviceReady
                    WriteToLog("Received: RD")

                    'Run the sequence
                    TrueTest.MDIForm.BeginInvoke(Sub() TrueTest.SequenceRunAll())

                    Exit Sub
                Case SerialCommand.PatternAcknowledge
                    WriteToLog("Received: PA," + ReceivedPacket.Data(0))
                    AckRecieved = True
                Case SerialCommand.UnloadAcknowledge
                    WriteToLog("Received: UA")
                Case SerialCommand.ResultAcknowledge
                    WriteToLog("Received: RA," + ReceivedPacket.Data(0))
                Case Else
                    WriteToLog("Unrecognized command: " & RecievedString)
            End Select

        Catch ex As Exception

            MessageBox.Show(ex.Message, "SerialPort Error")
            WriteToLog("Receive Error: " & ex.Message)

        End Try
    End Sub

#End Region

#Region "Event Handlers"

    Private Sub ExposureComplete(ByVal sender As Object, ByVal e As ExposureCompleteEventArgs)
        ExposuresCompleted += 1
        WriteToLog("Exposure Number " & ExposuresCompleted & " is complete.")
        If ExposuresCompleted >= NumberOfSequenceExposures Then
            'We've taken all of the exposures and we can unload the panel now
            Dim UnloadPacket As New PacketSerial(SerialCommand.Unload)
            UnloadPacket.Send(Me)
        End If
    End Sub

    Public Sub MeasurementComplete(ByVal sender As Object, ByVal e As MeasurementCompleteEventArgs)
        WriteToLog("Measurement Complete...")
    End Sub

    Public Sub AnalysisComplete(ByVal sender As Object, ByVal e As AnalysisCompleteEventArgs)
        WriteToLog("Analysis finished...")
    End Sub

    Public Sub SequenceComplete(ByVal sender As Object, ByVal e As SequenceCompleteEventsArgs)
        WriteToLog("Sequence Complete...")

        Dim ResultString As String = If(e.PassFail = TrueTest.AnalysisResultEnum.Pass, "OK", "NG")

        Dim DataStringList As New List(Of String) From {ResultString}

        Dim ResultPacket As New PacketSerial(SerialCommand.Result, DataStringList)
        ResultPacket.Send(Me)
        SequenceIsRunning = False
    End Sub

#End Region

    ''' <summary>
    ''' This is a blocking call that doesn't finish until the requested pattern is confirmed to be displayed.
    ''' </summary>
    ''' <param name="p"> The pattern to be displayed and confirmed. </param>
    Protected Overrides Sub ShowPattern(ByVal p As TrueTestPattern)
        If p Is Nothing Then Exit Sub

        'Check that the pattern is the appropriate type for the DI to send.
        If Not p.Pattern.GetType().Equals(GetType(SimplePattern)) Then Exit Sub

        AckRecieved = False
        Dim pattern As SimplePattern = CType(p.Pattern, SimplePattern)
        Dim PatternNumber As String = pattern.ImageNumber.ToString
        Dim PacketDataList As New List(Of String) From {PatternNumber}
        Dim PatternPacket As New PacketSerial(SerialCommand.ShowPattern, PacketDataList)
        PatternPacket.Send(Me)

        If Not WaitForAck() Then
            'Fixture timed out changing the pattern
            WriteToLog("Fixture did not send the Pattern Ready (PR) command after TrueTest requested pattern #" & PatternNumber & ".")
            TrueTest.SequenceStop() 'Cancel sequence
        End If

    End Sub

    Private Function WaitForAck() As Boolean
        'Wait for ack from board

        'loop in increments of 10ms
        Dim Timeout As Integer = WaitTime / 10
        Dim TimeoutCounter As Integer = 0
        Do
            System.Threading.Thread.Sleep(10)
            TimeoutCounter += 1
        Loop Until AckRecieved Or TimeoutCounter > Timeout
        If TimeoutCounter >= Timeout Then
            'never got anything back
            WriteToLog("No acknowledgement from board. Please check connection and power.")
        End If

        If AckRecieved Then
            AckRecieved = False
            Return True
        End If

        AckRecieved = False
        Return AckRecieved

    End Function

#Region "Log Files and Reports"
    ''' <summary>
    ''' Writes the string to the com log with a time stamp.
    ''' </summary>
    Private LogSyncLock As New Object
    Friend Sub WriteToLog(ByVal Message As String)
        Dim dt As Date = Now
        Dim LogFilename As String = LogFilePath & "\" & dt.ToString("yyyyMMdd") & " Comm Log.txt"
        Dim FileExists As Boolean = File.Exists(LogFilename)

        Try
            SyncLock LogSyncLock
                Dim sw As New StreamWriter(LogFilename, True)

                If FileExists = False Then
                    sw.WriteLine("Log created" & vbTab & dt.ToString("yyyy/MM/dd HH:mm:ss.fff"))
                End If

                If Message IsNot Nothing Then
                    Message = Message & vbTab & dt.ToString("yyyy/MM/dd HH:mm:ss.fff")
                    sw.WriteLine(Message)
                End If

                sw.Close()
            End SyncLock
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Log File Error")
        End Try
    End Sub

#End Region

    ''' <summary>
    ''' Checks the sequence for each unique measurement as analyses may reuse measurements already taken for others. 
    ''' </summary>
    Sub PrepareForSequenceRunAll(sender As Object, e As EventArgs)
        'Calculate how many exposures are going to be required to run the sequence
        Dim PatternList As New List(Of PatternSetup)
        NumberOfSequenceExposures = 0
        For Each SeqStep As TrueTestEngine.SequenceItem In TrueTest.Sequence.Items
            If Not SeqStep.Selected Then
                Continue For
            End If
            Dim ps As PatternSetup = TrueTest.Sequence.GetPatternSetupByName(SeqStep.PatternSetupName)
            If ps Is Nothing Then Continue For

            'Check to see if we already counted this pattern's exposures
            Dim RepeatPattern As Boolean = False
            For Each P As PatternSetup In PatternList
                If P = ps Then
                    RepeatPattern = True
                    Exit For
                End If
            Next

            If Not RepeatPattern Then
                PatternList.Add(ps)
                NumberOfSequenceExposures += 1
            End If
        Next
        WriteToLog("This sequence will require " & NumberOfSequenceExposures & " camera exposures.")
        'Reset the number of exposures completed
        ExposuresCompleted = 0
    End Sub
End Class
